pgxn/neon_walredo/walredoproc.c

```diff
@@ -226,7 +243,146 @@ WalRedoMain(int argc, char *argv[])
 	InitializeWalConsistencyChecking();
 #endif
 
-	CreateSharedMemoryAndSemaphores();
+	/*
+	 * Initialize dummy shmem.
+	 *
+	 * This code follows CreateSharedMemoryAndSemaphores() but manually sets up
+	 * the shmem header and skips few initializations that are not needed for
+	 * WAL redo.
+	 */
+	PGShmemHeader *shim = NULL;
+	{
```

Comment:
Let's put this to a separate function, `CreateFakeSharedMemoryAndSemaphores` or something, for clarity.

Reply:
Can we now revert the code changes we made in [src/backend/storage/ipc/ipci.c](https://github.com/neondatabase/postgres/blob/9a2093383ae19906f025b008ceecf89ebc9ea869/src/backend/port/sysv_shmem.c#L159)?

Reply:
Ah, I had a memory that we tried avoiding shmem in wal-redo, but missed that in the code. Probably all machinery that checks `am_wal_redo_postgres` and uses local buffers could be reverted too, will look into it.

Reply:
> Probably all machinery that checks am_wal_redo_postgres and uses local buffers could be reverted too, will look into it.

I think that was done to reduce the buffer locking overhead. Local buffers don't need locks.

Reply:
Ok, makes sense. Removed that malloc from `InternalIpcMemoryCreate()`.